### DISCUSSION TITLE
Run own Jinja2 templating and templates on pages

### DISCUSSION THREAD

**Discussion by Andre601** [mkdocs/mkdocs#3888] [Q&A]
I would like to setup my own Jinja2 templates to process through a Hook, but I'm unsure on what the best aproach would be.

The reason behind this is a current setup that I have, and that I'm not happy with.
The setup uses a custom template file, which reads page meta and generates page content from it using jinja2 macros.
This aproach has 2 problems tho:

- Templates are only processed after the markdown, so only raw HTML is supported for the page meta content to be rendered.
- MkDocs goes absolutely crazy about relative page links, as there aren't any headers yet that the link could point to, resulting in spamming the console.
  See #3885 for context.

Looking at solutions I have, running the pages with the meta through a own templater seems like a reasonable aproach, because I could execute it through the `on_page_markdown` method call, allowing me to inject markdown that could later on be processed further by MkDocs.
I already have such an aproach with creating a OpenAPI page from a JSON file ([Source](https://github.com/purrbot-site/Docs/blob/master/theme/.hooks/api_docs_gen.py)), where the templates contain Markdown instead of HTML for the most part, as this is done before the actual MD processing...

But there, I use a pre-made library that includes all the necessary Jinja2 setup. I'm not sure how I would recreate such a thing here for my use-case, especially if I would like to re-use pre-made filters by MkDocs alongside things like the page meta data...

And help here would be apreciated.

**Comment by kamilkrzyskow** [comment]
Hello @Andre601,
the macros plugin does render the pages in the [on_page_markdown event](https://github.com/fralau/mkdocs-macros-plugin/blob/e0ba1ba3683d5b618c979dd2d7bca5a48f7c87ad/mkdocs_macros/plugin.py#L899), and the `page` object together with `page.meta` is available. 

The issue could be getting other on_config data like config.site_name, but this could likely be hacked together as well. 
So I'm not sure how the plugin doesn't meet your expectations?

**Reply by Andre601** [reply]
As much as it may help (And as much as I used it before), there are a few core things I would need to first have cleared up before I could consider using it:

- Would there be a way to set jinja2 templating within a page template to use, so that it only is set once and I wouldn't need to paste the macro or code everywhere?
- This entire thing would be something I do not like to deal with, even with the python module's fix_url option:
  ![grafik](https://github.com/user-attachments/assets/1f71396b-43e0-4b66-8e2a-52aa328a455b)

**Reply by kamilkrzyskow** [reply]
> Would there be a way to set jinja2 templating within a page template to use, so that it only is set once and I wouldn't need to paste the macro or code everywhere?

I'm not sure what the question is, because I'm confused about the wording. Page templates in MkDocs can be set in the front matter header via `template: main.html` the default is `main.html` which extends from some other file like `base.html`. Those templates are part of MkDocs and are HTML-based, you can set a custom `page.meta.template` using a hook.

If you're planning to use another file as a Markdown Jinja template and include it using macros, and you don't want to input the `include` everywhere manually then you could use a hook with ~~lower~~ greater [event priority](https://www.mkdocs.org/dev-guide/plugins/#event-priorities) or use the [on_pre_page_macros](https://mkdocs-macros-plugin.readthedocs.io/en/latest/post_production/#application-to-headers-or-footers) event. This approach would still require some sort of "empty" Markdown files to be included in the event loop.

If you want no "physical" files and want to generate virtual files, then that's also possible, but I haven't done it myself.

> Relative links

A Markdown link to the same directory, should work `[title](samedirectoryfile.md)`, so the image seems to be wrong, unless there is something I'm not getting concerning the macros plugin 🤔 
Because the [directory urls](https://www.mkdocs.org/user-guide/configuration/#use_directory_urls) are the default, which you're probably using, HTML anchor links would require to use `<a href="../samedirectoryfile/index.html">title</a>`.

(EDIT: The image shows a link to `[other page](foo)`, however there is no file `foo` there is `foo.md`, the issue is the lack of `.md`)

This is something I've also explained in a previous discussion:
- https://github.com/mkdocs/mkdocs/discussions/3312


The solution for a dynamic template with moving paths would need some sort of filter like the [`url`](https://github.com/mkdocs/mkdocs/blob/7e4892ab2dd4a52efd95a9de407eee63310e0780/mkdocs/themes/mkdocs/base.html#L12) filter used in MkDocs, which does resolve HTML paths, so using a hook or macros filter you could expose a function which would convert a given string relative to the docs directory into a string relative to the current file based on the `page.file.src_path`.
- https://mkdocs-macros-plugin.readthedocs.io/en/latest/macros/#registration-of-variables-macros-and-filters

**Reply by Andre601** [reply]
> I'm not sure what the question is, because I'm confused about the wording. Page templates in MkDocs can be set in the front matter header via `template: main.html` the default is `main.html` which extends from some other file like `base.html`. Those templates are part of MkDocs and are HTML-based, you can set a custom `page.meta.template` using a hook.
> 
> If you're planning to use another file as a Markdown Jinja template and include it using macros, and you don't want to input the `include` everywhere manually then you could use a hook with lower [event priority](https://www.mkdocs.org/dev-guide/plugins/#event-priorities) or use the [on_pre_page_macros](https://mkdocs-macros-plugin.readthedocs.io/en/latest/post_production/#application-to-headers-or-footers) event. This approach would still require some sort of "empty" Markdown files to be included in the event loop.

I simply don't want to have to copy-paste a macro into every page there is... And preferably not use a macro at all as I am more confident with using template files with `if`s, `for` loops and similar.
The include thing, while also kinda annoying, would probably be an idea to consider here...

**Reply by Andre601** [reply]
Also, on an unrelated note, your docs are somewhat contradictory:

On the [Post-production page](https://mkdocs-macros-plugin.readthedocs.io/en/latest/post_production) you linked is it mentioning this at the start in an admonition box:

> ***From version 1.0.0***
> The `env.raw_markdown` attribute is deprecated. Use `env.markdown` instead.

But within the ["Application to headers or footers" section](https://mkdocs-macros-plugin.readthedocs.io/en/latest/post_production/#application-to-headers-or-footers) you linked is it mentioning the oposite:

> ***From version 1.0.0***
> `on_pre_page_macros(env)` works as shown only from that version.
>
> For lower versions, in `on_post_page_macros(env)` replace `env.raw_markdown` with `env.markdown`

Which is it now? Use `env.raw_markdown` or `env.markdown` if the version is lower?

And a final question is, if hooks do receive events from plugins, or if I have to do something such as import something...

If there is also some online reference for the API of the plugin would I apreciate a link, as I also wonder if I have access to stuff such as page meta through the `on_pre_page_macros` event, so that I only append data if present, or if I would need extra work there to be done to achieve this...

**Reply by Andre601** [reply]
Finding some old uses I had with this plugin, I slowly start to understand it again, but I encounter this nothing saying error on serve:
```
$ mkdocs serve --open
INFO    -  Building documentation...
INFO    -  [macros] - Macros arguments: {'module_name': 'docs/.hooks/api_docs', 'modules': [], 'render_by_default': True, 'include_dir': 'docs/.overrides', 'include_yaml': [], 'j2_block_start_string': '', 'j2_block_end_string': '', 'j2_variable_start_string': '', 'j2_variable_end_string': '', 'on_undefined': 'keep', 'on_error_fail': False, 'verbose': False}
INFO    -  [macros] - Found local Python module 'docs/.hooks/api_docs' in: D:\GitHub\AdvancedServerList
INFO    -  [macros] - Found external Python module 'docs/.hooks/api_docs' in: D:\GitHub\AdvancedServerList
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Scripts\mkdocs.exe\__main__.py", line 7, in <module>
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\click\core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\click\core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\click\core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\click\core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\click\core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\__main__.py", line 268, in serve_command
    serve.serve(**kwargs)
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\commands\serve.py", line 85, in serve
    builder(config)
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\commands\serve.py", line 67, in builder
    build(config, serve_url=None if is_clean else serve_url, dirty=is_dirty)
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\commands\build.py", line 265, in build
    config = config.plugins.on_config(config)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\plugins.py", line 587, in on_config
    return self.run_event('config', config)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs\plugins.py", line 566, in run_event
    result = method(item, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs_macros\plugin.py", line 585, in on_config
    self._load_modules()
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs_macros\plugin.py", line 458, in _load_modules
    self._load_module(module, local_module_name)
  File "C:\Users\Andreas\AppData\Local\Programs\Python\Python311\Lib\site-packages\mkdocs_macros\plugin.py", line 408, in _load_module
    raise NameError("No valid function found in module '%s'" %
NameError: No valid function found in module 'docs/.hooks/api_docs'
```
The macros plugin settings:
```yaml
plugins:
  # The plugin is last in the list
  - macros:
      include_dir: docs/.overrides
      module_name: docs/.hooks/api_docs
```
The `api_docs.py` file:
```py
def on_pre_page_macros(env):
    footer = "\n\n{% if page.meta and page.meta.api %}{% include 'api_docs/layout.md' %}{% endif %}"
    env.markdown += footer
```

**Reply by kamilkrzyskow** [reply]
> I simply don't want to have to copy-paste a macro into every page there is... And preferably not use a macro at all as I am more confident with using template files with ifs, for loops and similar.
> The include thing, while also kinda annoying, would probably be an idea to consider here...

I have no other ideas than this to make it more "general" / "automatic", as there needs to be some source of the instruction to template a given file. If not a hook, then the macros plugin would need to expose a config for that, and automatically include templates in the files, but I'm not aware of it. Also the macros part is the Jinja templating, so macros do include the ifs and for loops.

I only did some macros templates in the [mkdocs-nype](https://github.com/nypesap/mkdocs-nype/tree/main/mkdocs_nype/macros_includes) theme. Haven't used any of the importable modules etc. as pure MkDocs hooks / plugins were enough for me to modify the markdown.

> Which is it now? Use env.raw_markdown or env.markdown if the version is lower?

According to the changelog `env.markdown` in favour of `raw_markdown`
https://github.com/fralau/mkdocs-macros-plugin/blob/master/CHANGELOG.md#100-alpha-2023-04-23

> And a final question is, if hooks do receive events from plugins, or if I have to do something such as import something...

I don't understand the question, there is the MkDocs [event loop](https://www.mkdocs.org/dev-guide/plugins/#events) where both plugins and hooks register the event functions to be later executed in order according to their event priorities. As for the internal event loop of the macros plugin I haven't used it before.

> If there is also some online reference for the API of the plugin would I apreciate a link

Haven't found anything like the API docs you typically make. Just the typical reference docs.

> as I also wonder if I have access to stuff such as page meta through the on_pre_page_macros event

Also looking back, MkDocs did deprecate the `on_pre_page` event in favour of `on_page_markdown`, so perhaps `on_pre_page_macros` could also be replaced in full with `on_page_markdown`, and it does have access to all `markdown`, `page`, `config`.

Use a hook to inject Markdown / macros templating, before the macros plugin does render the Jinja on the page.

> but I encounter this nothing saying error on serve:

Given the traceback, and the error line, I see you're not using the `master`-ish latest version but perhaps 1.0.5
Based on the code it seems it looks for `define_env` or `define_variables` functions in the loaded module and can't find them.
Is the `docs/.hooks/api_docs` a file or a directory with submodules?
https://github.com/fralau/mkdocs-macros-plugin/blob/4bfda0bd7efe070439aa5cecea3d3ca2d86375d2/mkdocs_macros/plugin.py#L408

Without a runnable reproduction I can't debug what went wrong.

**Reply by Andre601** [reply]
> Also looking back, MkDocs did deprecate the `on_pre_page` event in favour of `on_page_markdown`, so perhaps `on_pre_page_macros` could also be replaced in full with `on_page_markdown`, and it does have access to all `markdown`, `page`, `config`.

That's not how it would work sadly.

Macros handles jinja tags before MkDocs does the pre_page_markdown call, so when I use that method, the jinja stuff won't be parsed anymore as it is past that moment, so I have to use their on_pre_page_macros call here...



> Given the traceback, and the error line, I see you're not using the `master`-ish latest version but perhaps 1.0.5
> Based on the code it seems it looks for `define_env` or `define_variables` functions in the loaded module and can't find them.
> Is the `docs/.hooks/api_docs` a file or a directory with submodules?
> https://github.com/fralau/mkdocs-macros-plugin/blob/4bfda0bd7efe070439aa5cecea3d3ca2d86375d2/mkdocs_macros/plugin.py#L408

Yeah, remember it now. It's that stupid issue I had in the past... And yes, the local version used was outdated. With the new one, this seems to now finally enable...

**Reply by Andre601** [reply]
Well, it finally inserts the stuff... Tho, now I face the main issue that it can't find another template I define in a file with `include`.

The include dir setting is the one I shared before and the location of the file is `<include_dir>/api_docs/parts/constructor_summaries.md`, yet when I use `{% include 'api_docs/parts/constructor_summaries.md' %}` the macros plugin raises a TemplateNotFound error.

**Reply by kamilkrzyskow** [reply]
Here is a minimal example, you didn't provide one, so I tried to keep it as close to the provided config as I understood:
- [macros-basic.zip](https://github.com/user-attachments/files/18264415/macros-basic.zip)

> That's not how it would work sadly.

That's how it works, perhaps you've not seen my fix in the message above, 
\>then you could use a hook with ~~lower~~ greater [event priority](https://www.mkdocs.org/dev-guide/plugins/#event-priorities)
I got mixed up and used lower instead of greater. Your hook event needs to run faster than the macros plugin's `on_page_markdown`, so a higher event priority is needed.

Even though personally, I would not keep any hooks or overrides in the `docs` directory this still works and I mimicked the path tree in the example above. Note that this `include_dir` path is relative to the `mkdocs serve` command, not relative to the config file location, but that's likely not the issue in your case, as it would otherwise not find the directory at all.

**Reply by Andre601** [reply]
You missed my comment where I mention that I managed to the the insertion working.
The only issue I face right now, is that the `{% include 'path/to/file' %}` doesn't work, as it can't find the provided file, despite the file being in the include_dir set for the plugin...

**Reply by kamilkrzyskow** [reply]
I did not miss this comment, I provided the zip which uses the example of `{% include 'api_docs/parts/constructor_summaries.md' %}`

Do the zipped files work for you, have you compared them to your setup? I don't know what could be the reason for the TemplateNotFound error on your end, maybe lack of permissions, maybe some edge case which I'm not seeing, misconfiguration or whatever, there is the [list_templates](https://jinja.palletsprojects.com/en/stable/api/#jinja2.Environment.list_templates) function of the Jinja Environment to see all available files that could be accessed.

Here I extended the above example to print out the templates:
- [macros-with-list-templates.zip](https://github.com/user-attachments/files/18266777/macros-with-list-templates.zip)

**Reply by Andre601** [reply]
`list_templates()` is not a function that seems to be included by default, and the Macros plugin seems to not be including it.

**Reply by kamilkrzyskow** [reply]
Happy New Year 🥳 

> list_templates() is not a function that seems to be included by default, and the Macros plugin seems to not be including it.

On what basis are you making this claim? The example above uses the `list_templates` function just fine 🤔 

```py
class Counter:
    count = 0

def on_pre_page_macros(env):
    if Counter.count > 0:
        return

    templates = env.env.list_templates()
    for t in templates:
        print("TEMPLATE:", t)

    Counter.count += 1
```

I added the primitive counter to execute the function only once, and not for every file.

I see there is the other discussion, so let's continue sharing examples there:
- https://github.com/fralau/mkdocs-macros-plugin/discussions/255

**Reply by Andre601** [reply]
Ah, you're using it within python code. I expected you meant using it within the template file/page itself.
